cfg_support/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+/* When cross compiling, build.rs runs on the host, thus when compiling it
+ * target = host. Use env vars listed in the links below for the actual target.
+ *
+ * https://doc.rust-lang.org/cargo/reference/environment-variables.html
+ * https://doc.rust-lang.org/reference/conditional-compilation.html
+ */
+
+#[derive(Debug)]
+pub struct CfgTarget {
+    pub target_arch: String,
+    pub target_feature: String,
+    pub target_os: String,
+    pub target_family: String,
+    pub target_env: String,
+    pub target_endian: String,
+    pub target_pointer_width: String,
+    pub target_vendor: String,
+}
+
+#[derive(Copy, Clone)]
+pub struct Cfg<'a> {
+    arch: Option<&'a str>,
+    // Not including target_feature here,
+    // it doesn't seem useful for an equality test.
+    //
+    // Usually you'll want to CfgTarget::new().target_feature.contains("....")
+    os: Option<&'a str>,
+    family: Option<&'a str>,
+    env: Option<&'a str>,
+    endian: Option<&'a str>,
+    pointer_width: Option<&'a str>,
+    vendor: Option<&'a str>,
+}
+
+#[derive(Copy, Clone)]
+pub enum CfgBuilder<'a> {
+    Uninitialized,
+    Cond(Cfg<'a>),
+}
+
+fn check_same(target: String, other: Option<&'_ str>) -> bool {
+    Some(target) == other.map(|x: &str| x.to_uppercase())
+}
+fn check_conflict(target: String, other: Option<&'_ str>) -> bool {
+    None == other || check_same(target, other)
+}
```

Comment:
I think there ought to be a way to implement this function such that `target` is also an `&str`, not a `String`, saving some allocations.

The function name `check_conflict` sounds to me like it should return `true` if there *is* a conflict. I would suggest renaming to `check_no_conflict`.

Reply:
Fixed the 2nd case, for the first case, I couldn't figure out a way, the issue is with
[std::str::to_uppercase](https://doc.rust-lang.org/std/primitive.str.html#method.to_uppercase)
We could avoid the copy by using: [std::str::make_ascii_uppercase](https://doc.rust-lang.org/std/primitive.str.html#method.make_ascii_uppercase)

If it's absolutely necessary I'll do so in a second pass, but I wanted to get the rest of your comments out of the way first.